litert/vendors/qualcomm/core/transformation/mha_to_sha.cc

```diff
@@ -0,0 +1,358 @@
+// Copyright (c) Qualcomm Innovation Center, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "litert/vendors/qualcomm/core/transformation/mha_to_sha.h"
+
+#include <array>
+#include <functional>
+#include <vector>
+
+#include "litert/vendors/qualcomm/core/builders/concatenation_op_builder.h"
+#include "litert/vendors/qualcomm/core/builders/reshape_op_builder.h"
+#include "litert/vendors/qualcomm/core/builders/slice_op_builder.h"
+#include "litert/vendors/qualcomm/core/builders/split_op_builder.h"
+#include "litert/vendors/qualcomm/core/op_code.h"
+#include "litert/vendors/qualcomm/core/tensor_pool.h"
+#include "litert/vendors/qualcomm/core/utils/log.h"
+#include "litert/vendors/qualcomm/core/wrappers/op_wrapper.h"
+#include "QnnInterface.h"  // from @qairt
+
+namespace qnn {
+namespace {
+
+// Returns a boolean indicating whether the output tensor of op1 at out_id is
+// connected to the input tensor of op2 at in_id.
+#define IS_CONNECTED(op1, out_id, op2, in_id)     \
+  (ops[start_id + op1].GetOutputTensor(out_id) == \
+   ops[start_id + op2].GetInputTensor(in_id))
+
+constexpr size_t kMulIndex = 0;
+constexpr size_t kTransposePrefillIndex = 1;
+constexpr size_t kReshapePrefillIndex = 2;
+constexpr size_t kMatMulK1Index = 1;
+constexpr size_t kMatMulK2Index = 2;
+constexpr size_t kConcatIndex = 3;
+constexpr size_t kReshape0Index = 4;
+constexpr size_t kAddIndex = 5;
+constexpr size_t kReshape1Index = 6;
+constexpr size_t kSoftmaxIndex = 7;
+constexpr size_t kSlice1Index = 8;
+constexpr size_t kSlice2Index = 9;
+constexpr size_t kMatMulV1Index = 10;
+constexpr size_t kMatMulV2Index = 11;
+constexpr size_t kAdd2Index = 12;
+constexpr size_t kReshape2Index = 13;
+constexpr size_t kTranspose2Index = 14;
+constexpr size_t kReshape3Index = 15;
+
+void CloneOpWithIO(
```

Comment:
According to the function, maybe this can be better.
```suggestion
void EmplaceOpWithIO(
```

Reply:
Done. Thanks!